vignettes/guides/making_new_layers_and_models_via_subclassing.Rmd

```diff
@@ -0,0 +1,969 @@
+---
+title: "Customizing what happens in `fit()`"
+author: "[fchollet](https://twitter.com/fchollet), [t-kalinowski](https://github.com/t-kalinowski)"
+output: rmarkdown::html_vignette
+date created: 2019/03/01
+last modified: 2020/04/13
+description: Complete guide to writing `Layer` and `Model` objects from scratch.
+---
+
+
+## Setup
+```{r, echo=FALSE}
+Sys.setenv(TF_CPP_MIN_LOG_LEVEL = 2)
+```
+
+```{r}
+library(magrittr)
+library(tensorflow)
+library(tfdatasets)
+library(keras)
+
+tf_version()
+```
+
+
+## The `Layer` class: the combination of state (weights) and some computation
+
+One of the central abstraction in Keras is the `Layer` class. A layer
+encapsulates both a state (the layer's "weights") and a transformation from
+inputs to outputs (a "call", the layer's forward pass).
+
+Here's a densely-connected layer. It has a state: the variables `w` and `b`.
+```{r}
+Linear(keras$layers$Layer) %py_class% {
+  initialize <- function(units = 32, input_dim = 32) {
+    super$initialize()
+    w_init <- tf$random_normal_initializer()
+    self$w <- tf$Variable(
+      initial_value = w_init(
+        shape = shape(input_dim, units),
+        dtype = "float32"
+      ),
+      trainable = TRUE
+    )
+    b_init <- tf$zeros_initializer()
+    self$b <- tf$Variable(
+      initial_value = b_init(shape = shape(units), dtype = "float32"),
+      trainable = TRUE
+    )
+  }
+
+  call <- function(inputs) {
+    tf$matmul(inputs, self$w) + self$b
+  }
+}
+```
+
+
+You would use a layer by calling it on some tensor input(s), much like a regular
+function.
+
+```{r}
+x <- tf$ones(shape(2, 2))
+linear_layer <- Linear(4, 2)
+y <- linear_layer(x)
+print(y)
+```
+
+All the builtin layers in the keras package (e.g., `layer_dense()`) are
```

Comment:
very nicely explained!

Reply:
on the other hand ... seeing this again some time later, now i think, - won't we always want this?
Do it transparently to the user?